cachito/web/status.py

```diff
@@ -0,0 +1,142 @@
+# SPDX-License-Identifier: GPL-3.0-or-later
+import requests
+from sqlalchemy.exc import SQLAlchemyError
+
+from cachito.errors import CachitoError
+from cachito.web import db
+from cachito.workers.config import get_worker_config
+from cachito.workers.tasks.celery import app
+
+
+no_retry_session = requests.Session()
+
+
+def _service_ok(url):
+    """
+    Check if a service is reachable and working correctly.
+
+    The status of the service is determined from the status code of the response.
+
+    :param str url: url to fetch
+    :return: tuple (ok: bool, reason: str or None)
+    """
+    try:
+        resp = no_retry_session.get(url)
+        resp.raise_for_status()
+    except requests.RequestException as e:
+        return False, str(e)
+
+    return True, None
+
+
+def nexus_ok(nexus_url):
+    """
+    Check if Nexus is OK.
+
+    Uses the /status endpoint, specifically /status/writable.
+
+    :param str nexus_url: url of the Nexus host
+    :return: tuple (ok: bool, reason: str or None)
+    """
+    nexus_status_url = f"{nexus_url.rstrip('/')}/service/rest/v1/status/writable"
+    return _service_ok(nexus_status_url)
+
+
+def athens_ok(athens_url):
+    """
+    Check if Athens is OK.
+
+    Simply checks if the root url is reachable (Athens has no status endpoint).
+
+    :param str athens_url: url of the Athens host
+    :return: tuple (ok: bool, reason: str or None)
+    """
+    return _service_ok(athens_url)
+
+
+def database_ok():
+    """
+    Check if the database connection is working.
+
+    :return: tuple (ok: bool, reason: str or None)
+    """
+    try:
+        # Does not actually query data, simply returns 1 (if the connection is working)
+        db.session().execute("SELECT 1")
+    except SQLAlchemyError as e:
+        return False, str(e)
+
+    return True, None
+
+
+def workers_status(*, short=False):
+    """
+    Ping workers, check received replies.
+
+    :param bool short: raise an error if a problem is found
+    :return: list of status information for individual workers (empty if no replies)
+    """
+    replies = app.control.inspect().ping() or {}
+    if short and not replies:
+        raise CachitoError("no workers responded to ping")
+
+    # replies is a dict of {hostname: reply}, convert to a sorted list of [(hostname, reply)]
+    reply_tuples = sorted(replies.items(), key=lambda kv: kv[0])
+    workers = []
+
+    for worker_name, reply in reply_tuples:
+        # The reply format does not seem to be documented anywhere, but is
+        # likely a dict with one key: "ok"/"error" and a corresponding message.
+        worker = {"name": worker_name, "ok": "ok" in reply}
+
+        if "ok" not in reply:
+            reason = reply.get("error", "unknown reason")
+            if short:
+                raise CachitoError(f"worker {worker_name} unavailable: {reason}")
```

Comment:
So, if 1 out of 12 workers is unavailable here, we'd respond with an error, but if 11 workers do not respond to the ping above and one do, we'd say everything is ok. Should we onlyreturn 503 if cachito is actually unable to process requests at the moment?

Would adding more information in a 200 response contents help here? Something like "all_systems_go: true/false"?
This of course would only be really truthful if we also included some sort of configuration to let Cachito know the number of workers expected to exist in a certain deployment (this last part may require further discussion).

Reply:
In this matter, should we also consider that Cachito may be partially unavailable? For instance, if Athens is unavailable, Cachito is still operational for python and npm requests.

Reply:
Hmm, all good points. I think I'll replace the `all_ok` attribute with something like `can_serve`, which will be True if all services and at least one worker are ok.

If we decide to add the configuration for expected number of workers, I'll also add `all_ok`.

I like the idea of partial availability, but I imagine that would best be left for another PR

Reply:
Done, including partial availability (only for /status, not /status/short)